src/hashupdater.cpp

```diff
@@ -5,19 +5,47 @@
  */
 
 #include "hashupdater.hpp"
-#include "util.hpp"
+#include "util/threadpool.hpp"
 #include "world.hpp"
 
 PasswordHashUpdater::PasswordHashUpdater(Config& config, const std::unordered_map<HashFunc, std::shared_ptr<Hasher>>& passwordHashers)
     : _config(config)
     , _passwordHashers(passwordHashers)
-    , _database(nullptr)
-    , _terminating(false)
-    , _updateSem(0)
 {
-    this->_updateThread = std::thread([this] { this->updateThreadProc(); });
+}
+
+void PasswordHashUpdater::QueueUpdatePassword(const std::string& username, util::secure_string&& password, HashFunc hashFunc)
+{
+    auto updateThreadProc = [this](const void * state)
+    {
+        auto updateState = reinterpret_cast<const PasswordHashUpdater::UpdateState*>(state);
+
+        auto username = updateState->username;
+        auto updatedPassword = std::move(Hasher::SaltPassword(std::string(this->_config["PasswordSalt"]),
+                                                              username,
+                                                              std::move(const_cast<PasswordHashUpdater::UpdateState*>(updateState)->password)));
+        auto hashFunc = updateState->hashFunc;
+
+        if (hashFunc == NONE)
+            return;
+
+        updatedPassword = std::move(this->_passwordHashers[hashFunc]->hash(std::move(updatedPassword.str())));
+
+        this->CreateDbConnection()->Query("UPDATE `accounts` SET `password` = '$', `password_version` = # WHERE `username` = '$'",
+            updatedPassword.str().c_str(),
+            hashFunc,
+            username.c_str());
+
+        delete updateState;
+    };
 
-    auto dbType = util::lowercase(std::string(config["DBType"]));
+    auto state = reinterpret_cast<void*>(new UpdateState { username, std::move(password), hashFunc });
```

Comment:
I've always just kind of assumed `new` won't fail... :)

Assuming `new` does fail, I think there will be bigger problems with eoserv, yeah? Isn't a failure of `new` usually an out of memory scenario?

Reply:
I actually realized I've never written code to handle failure of `new`. What's the common approach to take for this? Just add a `(std::nothrow)` to the call?

Reply:
I ended up just adding `std::nothrow`. I can change this if there's a better thing to do but this seems to be the standard way in modern C++

Reply:
> I've always just kind of assumed `new` won't fail

Yeah, it's definitely an edge case and I only really see it when working at the limits of embedded devices. When I wrote the comment I was questioning if it was worth addressing here. I decided to comment on it since (a) users might deploy it on a low cost server with restricted resources (e.g., 10G HDD, 512MB RAM) and (b) since this is for a MMO game it is possible that a high spike in players could consume all the resources on any device.

> I ended up just adding `std::nothrow`

That's the way I would have done it as well